go/src/koding/kites/kloud/provider/builder.go

```diff
@@ -311,13 +311,33 @@ func (bs *BaseStack) applyAsync(ctx context.Context, req *kloud.ApplyRequest) er
 		return err
 	}
 
+	// At this point resources have been built successfully on the terraform side.
+	// If updating stack or provisioning klient on remote hosts fails,
+	// we need to cleanup the stack; if debug: true was set in the stack template
+	// for any of the instance resources, the stack won't be destroyed
+	// so it's possible to troubleshoot it post-mortem.
+	if !bs.Debug {
+		defer func() {
+			if err != nil {
+				_, e := tfKite.Destroy(&tf.TerraformRequest{
```

Comment:
> At this point I am not sure whether not updating state file on failure was intentional design for terraformer or not. 

It was not intentional AFAIR.

Just checked the code again and after this [commit](https://github.com/koding/koding/commit/3c9ae8c3e0cd7a20a7319d8daa67dd5f990f6240#diff-5ce92a83686bcbe1f62e027cee0f5f82L30) we broke that support.

Do you see any problem on adding it back to system? Could we solve the problem with a proper error message and some UI operations? (eg: your stack build failed but you can retry building it - build button chages to re-build and we can add destroy button next to it?) <- this part is just a thought we can talk with product team